Stratis.Bitcoin.Features.LightWallet/LightWalletSyncManager.cs

```diff
@@ -110,43 +110,79 @@ public void Stop()
         public void ProcessBlock(Block block)
         {
-            // if the new block previous hash is the same as the 
-            // wallet hash then just pass the block to the manager 
+            Guard.NotNull(block, nameof(block));
+            this.logger.LogTrace("({0}:'{1}')", nameof(block), block.GetHash());
+
+            var newTip = this.chain.GetBlock(block.GetHash());
+            if (newTip == null)
+            {
+                this.logger.LogTrace("(-)[NEW_TIP_REORG]");
+                return;
+            }
+
+            // If the new block's previous hash is the same as the 
+            // wallet hash then just pass the block to the manager. 
             if (block.Header.HashPrevBlock != this.walletTip.HashBlock)
             {
-                // if previous block does not match there might have 
-                // been a reorg, check if the wallet is still on the main chain
+                // If previous block does not match there might have 
+                // been a reorg, check if the wallet is still on the main chain.
                 ChainedBlock inBestChain = this.chain.GetBlock(this.walletTip.HashBlock);
                 if (inBestChain == null)
                 {
-                    // the current wallet hash was not found on the main chain
-                    // a reorg happenend so bring the wallet back top the last known fork
+                    // The current wallet hash was not found on the main chain.
+                    // A reorg happened so bring the wallet back top the last known fork.
 
                     ChainedBlock fork = this.walletTip;
 
-                    // we walk back the chained block object to find the fork
+                    // We walk back the chained block object to find the fork.
                     while (this.chain.GetBlock(fork.HashBlock) == null)
                         fork = fork.Previous;
 
-                    Guard.Assert(fork.HashBlock == block.Header.HashPrevBlock);
+                    this.logger.LogInformation("Reorg detected, going back from '{0}/{1}' to '{2}/{3}'.", this.walletTip.HashBlock, this.walletTip.Height, fork.HashBlock, fork.Height);
+
                     this.walletManager.RemoveBlocks(fork);
-                    this.logger.LogWarning($"Reorg detected, wallet tip reverted back to Height = {fork.Height} hash = {fork.HashBlock}.");
+                    this.walletTip = fork;
+
+                    this.logger.LogTrace("Wallet tip set to '{0}'.", this.walletTip);
+                }
+
+                // The new tip can be ahead or behind the wallet
+                // If the new tip is ahead we try to bring the wallet up to the new tip.
+                // If the new tip is behind we just check the wallet and the tip are in the same chain.
+
+                if (newTip.Height > this.walletTip.Height)
+                {
+                    ChainedBlock findTip = newTip.FindAncestorOrSelf(this.walletTip.HashBlock);
```

Comment:
That's a great idea actually, this can also stop at the last checkpoint (we will need to put the checkpoint on the network class for that.

I will add an issue to do that.